### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by namitmax**

### CODE REVIEW

File: include/afina/concurrency/Executor.h [hunk 1]
```diff
@@ -28,16 +32,44 @@ class Executor {
...
      * Signal thread pool to stop, it will stop accepting new jobs and close threads just after each become
      * free. All enqueued jobs will be complete.
      *
      * In case if await flag is true, call won't return until all background jobs are done and all threads are stopped
      */
-    void Stop(bool await = false);
+    void Stop(bool await = false) {
+        state = State::kStopping;
+        {
+            std::unique_lock<std::mutex> condition_lock(mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Не нужно

File: include/afina/concurrency/Executor.h [hunk 2]
```diff
@@ -28,16 +32,44 @@ class Executor {
...
+            empty_condition.notify_all();
+        }
+        for (size_t i = 0; i < threads.size(); ++i) {
+            if (await) {
+                threads[i].join();
+            }
+        }
+        state = State::kStopped;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Выставляем только если тредпул завершен

File: include/afina/concurrency/Executor.h [hunk 3]
```diff
@@ -28,16 +32,44 @@ class Executor {
...
 
     /**
      * Signal thread pool to stop, it will stop accepting new jobs and close threads just after each become
      * free. All enqueued jobs will be complete.
      *
      * In case if await flag is true, call won't return until all background jobs are done and all threads are stopped
      */
-    void Stop(bool await = false);
+    void Stop(bool await = false) {
+        state = State::kStopping;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Защитить мьютексом

File: include/afina/concurrency/Executor.h [hunk 4]
```diff
@@ -50,12 +82,23 @@ class Executor {
         // Prepare "task"
         auto exec = std::bind(std::forward<F>(func), std::forward<Types>(args)...);
 
-        std::unique_lock<std::mutex> lock(this->mutex);
-        if (state != State::kRun) {
+        if (state != State::kRun || _current_size >= max_queue_size) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Защитить мьютексом

File: include/afina/concurrency/Executor.h [hunk 5]
```diff
@@ -50,12 +82,23 @@ class Executor {
...
             return false;
         }
 
         // Enqueue new task
+        {
+            std::unique_lock<std::mutex> lock(mutex);
+            if (threads.size() < high_watermark
+                && _current_size++ > 0
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Следующее условие может оказаться ложным

Comment (Fen99) [reply include/afina/concurrency/Executor.h]: Новое задание точно будет

File: include/afina/concurrency/Executor.h [hunk 6]
```diff
@@ -50,12 +82,23 @@ class Executor {
...
+                && threads.size() == _working) {
+                threads.emplace_back(std::thread([this] {
+                    return perform(this);
+                }));
+            }
+        }
+        std::unique_lock<std::mutex> lock(this->mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Лучше на одном мьютексе делать

File: include/afina/concurrency/Executor.h [hunk 7]
```diff
@@ -71,7 +114,51 @@ class Executor {
     /**
      * Main function that all pool threads are running. It polls internal task queue and execute tasks
      */
-    friend void perform(Executor *executor);
+    friend void perform(Executor *executor) {
+        auto now = std::chrono::system_clock::now();
+        while (executor->state == Executor::State::kRun) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Под мьютексом

File: include/afina/concurrency/Executor.h [hunk 8]
```diff
@@ -71,7 +114,51 @@ class Executor {
...
+                            condition_lock,
+                            now + std::chrono::milliseconds(executor->_time));
+                        if (executor->state != Executor::State::kRun ||
+                                executor->tasks.empty() &&
+                                executor->threads.size() > executor->low_watermark &&
+                                wake == std::cv_status::timeout) {
+                            if (executor->state != Executor::State::kStopping) {
+                                std::unique_lock<std::mutex> lock(executor->_mutex);
+                                auto id = std::this_thread::get_id();
+                                auto find_it =
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: vector -> map

File: include/afina/concurrency/Executor.h [hunk 9]
```diff
@@ -28,16 +32,44 @@ class Executor {
...
-    void Stop(bool await = false);
+    void Stop(bool await = false) {
+        state = State::kStopping;
+        {
+            std::unique_lock<std::mutex> condition_lock(mutex);
+            empty_condition.notify_all();
+        }
+        for (size_t i = 0; i < threads.size(); ++i) {
+            if (await) {
+                threads[i].join();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: if (joinable)

File: include/afina/concurrency/Executor.h [hunk 10]
```diff
@@ -71,7 +114,51 @@ class Executor {
...
+                                                     return thread.get_id() == id;
+                                                 });
+                                executor->threads.erase(find_it);
+                            }
+                            return;
+                        }
+                    }
+                    work_func = std::move(executor->tasks.front());
+                    executor->tasks.pop_front();
+                    executor->_current_size--;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Зачем _current_size, если есть tasks.size() ?

File: include/afina/concurrency/Executor.h [hunk 11]
```diff
@@ -71,7 +114,51 @@ class Executor {
...
+                    executor->_working++;
+                }
+            }
+            try {
+                work_func();
+            } catch (...) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:151]: catch(const std::exception&) - вдруг можно залогировать что-то осмысленное?

**Review by Fen99 — CHANGES_REQUESTED** [review]
~ServerImpl = Stop() + Join()

File: include/afina/concurrency/Executor.h [hunk 12]
```diff
@@ -71,7 +114,43 @@ class Executor {
...
+                    std::unique_lock<std::mutex> condition_lock(executor->mutex);
+                    while (executor->tasks.empty()) {
+                        auto wake = executor->empty_condition.wait_until(
+                            condition_lock, now + std::chrono::milliseconds(executor->_time));
+                        if (executor->state != Executor::State::kRun ||
+                            executor->tasks.empty() && executor->threads.size() > executor->low_watermark &&
+                                wake == std::cv_status::timeout) {
+                            if (executor->state != Executor::State::kStopping) {
+                                auto id = std::this_thread::get_id();
+                                executor->threads.erase(id);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Сначала detach()

File: include/afina/concurrency/Executor.h [hunk 13]
```diff
@@ -28,16 +33,47 @@ class Executor {
...
+                            low_watermark(min),
+                            high_watermark(max),
+                            _time(time) {
+             std::unique_lock<std::mutex> lock(mutex);
+             for (size_t  i = 0; i < low_watermark; ++i) {
+                 std::thread thread = std::thread([this] {
+                     return perform(this);
+                 });
+                 threads.emplace(thread.get_id(), std::move(thread));
+             }
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:49]: state = kRun

File: include/afina/concurrency/Executor.h [hunk 14]
```diff
@@ -28,16 +33,47 @@ class Executor {
...
      */
-    void Stop(bool await = false);
+    void Stop(bool await = false) {
+        {
+            std::unique_lock<std::mutex> _lock(mutex);
+            state = State::kStopping;
+        }
+        empty_condition.notify_all();
+        for (auto &thread : threads) {
+            if (await && thread.second.joinable()) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Переписать на conditional_variable

File: include/afina/concurrency/Executor.h [hunk 15]
```diff
@@ -71,7 +114,43 @@ class Executor {
     /**
      * Main function that all pool threads are running. It polls internal task queue and execute tasks
      */
-    friend void perform(Executor *executor);
+    friend void perform(Executor *executor) {
+        auto now = std::chrono::system_clock::now();
+        {
+            std::unique_lock<std::mutex> lock(executor->_mutex);
+            while (executor->state == Executor::State::kRun) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: || !tasks.empty()

File: include/afina/concurrency/Executor.h [hunk 16]
```diff
@@ -71,7 +114,43 @@ class Executor {
...
+                std::function<void()> work_func;
+                {
+                    std::unique_lock<std::mutex> condition_lock(executor->mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: 1) now нужно засекать здесь
2) второй раз берется тот же мьютекс

File: include/afina/concurrency/Executor.h [hunk 17]
```diff
@@ -71,7 +114,43 @@ class Executor {
...
+                    }
+                    work_func = std::move(executor->tasks.front());
+                    executor->tasks.pop_front();
+                    executor->_working++;
+                }
+                try {
+                    work_func();
+                } catch (const std::exception &err) {
+                    std::cout << "Some error  occurred during the  function execution, what : " << err.what();
+                }
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: catch(...)

File: include/afina/concurrency/Executor.h [hunk 18]
```diff
@@ -71,7 +114,43 @@ class Executor {
...
+                            }
+                            return;
+                        }
+                    }
+                    work_func = std::move(executor->tasks.front());
+                    executor->tasks.pop_front();
+                    executor->_working++;
+                }
+                try {
+                    work_func();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Таска исполняется не под мьютексом

File: include/afina/concurrency/Executor.h [hunk 19]
```diff
@@ -71,7 +114,43 @@ class Executor {
...
+                } catch (const std::exception &err) {
+                    std::cout << "Some error  occurred during the  function execution, what : " << err.what();
+                }
+                {
+                    std::unique_lock<std::mutex> lock(executor->mutex);
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Опять тот же мьютекс

File: include/afina/concurrency/Executor.h [hunk 20]
```diff
@@ -86,8 +165,8 @@ class Executor {
     /**
      * Vector of actual threads that perorm execution
      */
-    std::vector<std::thread> threads;
-
+    std::unordered_map<std::thread::id, std::thread> threads;
+    std::vector<std::thread> threads_ids;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Не нужен

**Review by Fen99 — CHANGES_REQUESTED** [review]

File: include/afina/concurrency/Executor.h [hunk 21]
```diff
@@ -28,16 +33,57 @@ class Executor {
...
+            }
+        }*/
+        {
+            std::unique_lock<std::mutex> _lock(mutex);
+            if (await) {
+                while (!threads.empty()){
+                    stop_condition.wait(_lock);
+                }
+            }
+            state = State::kStopped;
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: kStopped здесь мы выставляем только если working == 0, иначе должен выставлять последний завершающийся поток

File: include/afina/concurrency/Executor.h [hunk 22]
```diff
@@ -71,7 +124,49 @@ class Executor {
     /**
      * Main function that all pool threads are running. It polls internal task queue and execute tasks
      */
-    friend void perform(Executor *executor);
+    friend void perform(Executor *executor) {
+        std::unique_lock<std::mutex> lock(executor->_mutex);
+        while (executor->state == Executor::State::kRun || !executor->tasks.empty()) {
+            std::function<void()> work_func;
+            while (executor->tasks.empty()) {
+                auto now = std::chrono::system_clock::now();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Не решена проблема с бесконечным ожиданием

File: include/afina/concurrency/Executor.h [hunk 23]
```diff
@@ -71,7 +124,49 @@ class Executor {
...
+                auto wake = executor->empty_condition.wait_until(
+                            lock, now + std::chrono::milliseconds(executor->_time));
+                if (executor->state != Executor::State::kRun ||
+                    executor->tasks.empty() && executor->threads.size() > executor->low_watermark &&
+                        wake == std::cv_status::timeout) {
+                    if (executor->state != Executor::State::kStopping) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: join уже нет, можно удалять

File: include/afina/concurrency/Executor.h [hunk 24]
```diff
@@ -71,7 +124,49 @@ class Executor {
...
+                        auto id = std::this_thread::get_id();
+                        executor->threads.at(id).detach();
+                        executor->threads.erase(id);
+                    }
+                    return;
+                }
+                if (executor->state != Executor::State::kRun) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Недостижимый код
!kRun && tasks.empty()

File: include/afina/concurrency/Executor.h (hunk 23 above)
Comment (Fen99) [reply include/afina/concurrency/Executor.h]: Перенести код удаления треда и notify за общий цикл, в конец функции, здесь использовать break

File: include/afina/concurrency/Executor.h [hunk 25]
```diff
@@ -71,7 +124,49 @@ class Executor {
...
+                    auto id = std::this_thread::get_id();
+                    executor->threads.at(id).detach();
+                    executor->threads.erase(id);
+                    if (executor->threads.empty()) {
+                        executor->stop_condition.notify_all();
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h]: Выставить kStopped

File: include/afina/concurrency/Executor.h [hunk 26]
```diff
@@ -28,16 +33,57 @@ class Executor {
...
 
     /**
      * Signal thread pool to stop, it will stop accepting new jobs and close threads just after each become
      * free. All enqueued jobs will be complete.
      *
      * In case if await flag is true, call won't return until all background jobs are done and all threads are stopped
      */
-    void Stop(bool await = false);
+    void Stop(bool await = false) {
+        {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:67]: if (kStopped) { return; }

**Review by Fen99 — CHANGES_REQUESTED** [review]

File: include/afina/concurrency/Executor.h [hunk 27]
```diff
@@ -71,7 +121,47 @@ class Executor {
...
+                auto wake = executor->empty_condition.wait_until(
+                            lock, now + std::chrono::milliseconds(executor->_time));
+                if (executor->state != Executor::State::kRun ||
+                    executor->tasks.empty() && executor->threads.size() > executor->low_watermark &&
+                        wake == std::cv_status::timeout) {
+                    loop_stat = true;
+                    break;
+                }
+            }
+            if (loop_stat) {
```
Comment (Fen99) [inline include/afina/concurrency/Executor.h:140]: && !tasks.empty()

**Review by Fen99 — APPROVED** [review]
С учетом небольших изменений
